server/catgenome/src/main/java/com/epam/catgenome/util/feature/reader/TabixFeatureReader.java

```diff
@@ -0,0 +1,253 @@
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+import htsjdk.samtools.util.BlockCompressedInputStream;
+import htsjdk.samtools.util.RuntimeIOException;
+import htsjdk.tribble.AsciiFeatureCodec;
+import htsjdk.tribble.CloseableTribbleIterator;
+import htsjdk.tribble.Feature;
+import htsjdk.tribble.TribbleException;
+import htsjdk.tribble.readers.LineReader;
+import htsjdk.tribble.readers.LineReaderUtil;
+import htsjdk.tribble.readers.PositionalBufferedStream;
+import htsjdk.tribble.util.ParsingUtils;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+
+/**
+ * @author Jim Robinson
+ * @since 2/11/12
+ */
+public class TabixFeatureReader<T extends Feature, S> extends AbstractFeatureReader<T, S> {
+
+    TabixReader tabixReader;
+    List<String> sequenceNames;
+    String indexFile;
+
+    /**
+     *
+     * @param featureFile - path to a feature file. Can be a local file, http url, or ftp url
+     * @param codec
+     * @throws IOException
+     */
+    public TabixFeatureReader(final String featureFile, final AsciiFeatureCodec codec) throws IOException {
+        super(featureFile, codec);
+        tabixReader = new TabixReader(featureFile);
+        sequenceNames = new ArrayList<>(tabixReader.getChromosomes());
+        readHeader();
+    }
+
+    /**
+     *
+     *
+     * @param featureFile - path to a feature file. Can be a local file, http url, or ftp url
+     * @param indexFile - path to the index file.
+     * @param codec
+     * @throws IOException
+     */
+    public TabixFeatureReader(final String featureFile, final String indexFile,
+                              final AsciiFeatureCodec codec, EhCacheBasedIndexCache indexCache) throws IOException {
+        super(featureFile, codec);
+
+        this.indexCache = indexCache;
+        this.indexFile = indexFile;
+        tabixReader = new TabixReader(featureFile, indexFile, indexCache);
+        sequenceNames = new ArrayList<>(tabixReader.getChromosomes());
+
+        readHeader();
+    }
+
+    /**
+     * read the header
+     *
+     * @return a Object, representing the file header, if available
+     * @throws IOException throws an IOException if we can't open the file
+     */
+    private void readHeader() throws IOException {
+        S source = null;
+        String[] indexFileSplit = indexFile.split("\\?");
```

Comment:
Not sure if this regular expression is okay... Looks like it just splits every symbol. Plus, I guess it would be better to have the full path as a key, to make sure thet there won't be collisions on equal file names.